Blaze: Fix state change warnings when editing ads

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #11780 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
In SwiftUI, if we try to change a view's state just before it's about to change, we get the warning "Publishing changes from within view updates is not allowed".

In `BlazeEditAdViewModel`, we violated this in two places:
- We observe `tagline` to change `taglineRemainingLength` immediately. Please note, we get a signal for change of `tagline` just before the value is changed. When we change `taglineRemainingLength` in `sink`, we are triggering an update to the SwiftUI view that listens to the property, before the the view even updates following the change to `tagline`.
- On `BlazeEditAdView`, we observe `tagline` with `onChange` and trigger the update to the value if the text exceeds the character limit. This again triggers state change even before the view is updated for `tagline`.


Solution:
- Updated `taglineRemainingLength` and `taglineEmptyError` using `map` instead of `sink` when observing `tagline`. 
- Added a similar fix for `descriptionRemainingLength` and `descriptionEmptyError`.
- Removed the forced formatting of tagline and description. Instead, we should let the user continue typing, while displaying the character count in red to signal an error.


## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
- Build the app and log in to a store eligible for Blaze.
- Select Promote in the Blaze secion on the Dashboard screen.
- Select Edit Ad to change the content.
- Type in the tagline and description fields until you exceed the character limits. Confirm the following:
  - The text should be displayed as is
  - The character count should be red when the text length exceeds the limit.
  - No warnings on Xcode regarding `Publishing changes from within view updates is not allowed`.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/woocommerce/woocommerce-ios/assets/5533851/071cb827-f5b7-49d5-ae18-32c664367dde



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

itsmeichigo committed:
Update logic for updating descriptionRemainingLength

itsmeichigo committed:
Remove forced state change when editing text

dangermattic commented:
<!--
  0 Errors
  2 Warnings: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/182">17.1</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Blaze/BlazeEditAd/BlazeEditAdViewModel.swift:
```diff
@@ -21,33 +21,37 @@ final class BlazeEditAdViewModel: ObservableObject {
     var isSaveButtonEnabled: Bool {
-        guard let editedAdData else {
+        guard let editedAdData,
+                taglineRemainingLength >= 0,
+                descriptionRemainingLength >= 0 else {
```
nit: Kinldly consider adding unit tests for this.

itsmeichigo replied:
Added tests in 444424c9cdfe5eefe3b876072ff89801007207ee.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Blaze/BlazeEditAd/BlazeEditAdView.swift:
```diff
@@ -118,6 +115,7 @@ private extension BlazeEditAdView {
             )
 
             Text(viewModel.taglineFooterText)
+                .foregroundStyle(viewModel.taglineRemainingLength >= 0 ? .secondary : Color(.error))
```
Nice work using the `error` color. 

Should we use `.error` color also when the tagline is empty? Applies to the description field as well. 

<img src="https://github.com/woocommerce/woocommerce-ios/assets/524475/3538a818-9e07-4f6a-8074-f8b36f98aa5c" width="320"/>

itsmeichigo replied:
Good point - I added properties for validating the fields and updated the view to use them in 34aaf628f909d6b2561f8adcb6c78e7792b867b1. Test added in 053d536e370b8bd1184126a70b7407f7298beb1b.

selanthiraiyan reviewed (approved):
Thanks for the fix and for explaining in detail the reason and solution. 🙇

itsmeichigo committed:
Update unit tests for save buttons in edit ad view

itsmeichigo committed:
Show footer in red if tagline or description is empty

itsmeichigo committed:
Clean up logic for taglineFooterText and descriptionFooterText

itsmeichigo committed:
Add tests for isTaglineValidated and isDescriptionValidated

itsmeichigo committed:
Remove unused variable and import

itsmeichigo committed:
Update more friendly error message for exceeding character limit

itsmeichigo committed:
Merge branch 'trunk' into fix/11780-fix-warning

itsmeichigo committed:
Update tests for tagline and description footers

itsmeichigo commented:
Thank you, @selanthiraiyan, for the reviews! I went ahead and addressed the issues you mentioned in the comment in the subsequent PRs. Also, I realized that the messages for remaining characters with a minus number are not user friendly, so I also added a change to use a generic error message in 102dbb3e756514dae8ad43fe3c69bdf8bd8a4cc3 (with tests in 9e5214bd5e06d39d4340de7318b56f40b3728055). The messages should now be like this:

<img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/a6334578-4238-4bda-96bc-4bf93a75cb74" width=320 />

I'll enable auto-merge for now - please feel free to leave any comments and I'll address them in another PR if needed.
